django/thunderstore/api/cyberstorm/views/package_preview.py

```diff
@@ -0,0 +1,288 @@
+
+        if params.get("user_identifier", None):
+            filter_q.add(
+                Q(
+                    package__owner__members__user__username__icontains=params[
+                        "user_identifier"
+                    ]
+                ),
+                Q.AND,
+            )
+
+        if params.get("package_identifier", None):
+            filter_q.add(Q(package__name=params["package_identifier"]), Q.AND)
+
+        if not params.get("deprecated", None):
+            exclude_q.add(Q(package__is_deprecated=True), Q.OR)
+
+        if not params.get("nsfw", None):
+            exclude_q.add(Q(has_nsfw_content=True), Q.OR)
+
+        if params.get("q", None):
+            icontains_query = Q()
+            parts = [x for x in params["q"].split(" ") if x]
+
+            for part in parts:
+                for field in (
+                    "package__name",
+                    "package__owner__name",
+                    "package__latest__description",
+                ):
+                    icontains_query &= ~Q(**{f"{field}__icontains": part})
+            exclude_q.add(icontains_query, Q.OR)
+
+        # Add category filters/excludes
+        included = []
+        excluded = []
+        if params.get("section", None):
+            try:
+                section = PackageListingSection.objects.prefetch_related(
+                    "require_categories", "exclude_categories"
+                ).get(slug=params["section"])
+            except PackageListingSection.DoesNotExist:
+                pass
+            else:
+                [included.append(s.slug) for s in section.require_categories.all()]
+                [excluded.append(s.slug) for s in section.exclude_categories.all()]
+
+        if params.get("included_categories", None):
+            included = list(chain(included, params["included_categories"]))
+
+        if params.get("excluded_categories", None):
+            excluded = list(chain(excluded, params["excluded_categories"]))
+
+        if len(included) > 0:
+            included_cats_q = Q()
+            for cat in sorted(set(included)):
+                included_cats_q.add(Q(categories__slug__icontains=cat), Q.OR)
+            filter_q.add(included_cats_q, Q.AND)
+
+        if len(excluded) > 0:
+            excluded_cats_q = Q()
+            for cat in sorted(set(excluded)):
+                excluded_cats_q.add(Q(categories__slug__icontains=cat), Q.OR)
+            exclude_q.add(excluded_cats_q, Q.OR)
+
+        return (
+            PackageListing.objects.active()
+            .filter(
+                review_status__icontained_by=Case(
+                    When(
+                        Q(community__require_package_listing_approval=True),
+                        then=Value("approved"),
+                    ),
+                    default=Value("unreviewed approved"),
+                )
+            )
+            .filter(filter_q)
+            .exclude(exclude_q)
```

Comment:
Chained filters can lead to unintended consequences. Chaining two filters together gives you results that meet either criteria (at least in some instances), rather than only results that meet both criteria. Chain excludes instead or only use a single filter call with all filter criteria included in the same one (by composing Q objects).

Reply:
Alright, that's weird. I've now added the first filter as an Q object in the second filter.